src/Agent.Listener/Configuration/ConfigurationProvider.cs

```diff
@@ -367,4 +366,176 @@ public override async Task GetPoolIdAndName(AgentSettings agentSettings, Command
             }
         }
     }
+
+    public class EnvironmentVMResourceConfigProvider : DeploymentGroupAgentConfigProvider, IConfigurationProvider
+    {
+        public new string ConfigurationProviderType
+            => Constants.Agent.AgentConfigurationProvider.EnvironmentVMResourceConfiguration;
+        private IEnvironmentsServer _environmentsServer = null;
+
+        public override void Initialize(IHostContext hostContext)
+        {
+            base.Initialize(hostContext);
+            _environmentsServer = HostContext.GetService<IEnvironmentsServer>();
+        }
+
+        public override async Task TestConnectionAsync(AgentSettings agentSettings, VssCredentials creds, bool isHosted)
+        {
+            var url = agentSettings.ServerUrl;  // Ensure not to update back the url with agentSettings !!!
+            _term.WriteLine(StringUtil.Loc("ConnectingToServer"));
+
+            // Create the connection for environment virtual machine resource
+            Trace.Info("Test connection with environment");
+            if (!isHosted && !string.IsNullOrWhiteSpace(agentSettings.CollectionName)) // For on-prm validate the collection by making the connection
+            {
+                UriBuilder uriBuilder = new UriBuilder(new Uri(url));
+                uriBuilder.Path = uriBuilder.Path + "/" + agentSettings.CollectionName;
+                Trace.Info("Tfs Collection level url to connect - {0}", uriBuilder.Uri.AbsoluteUri);
+                url = uriBuilder.Uri.AbsoluteUri;
+            }
+            VssConnection environmentConnection = VssUtil.CreateConnection(new Uri(url), creds);
+
+            await _environmentsServer.ConnectAsync(environmentConnection);
+            Trace.Info("Connection complete for environment");
+        }
+
+        public override async Task GetPoolIdAndName(AgentSettings agentSettings, CommandSettings command)
+        {
+            _projectName = command.GetProjectName(_projectName);
+            var environmentName = command.GetEnvironmentName();
+            Trace.Info("vm resource will be configured against the environment '{0}'", environmentName);
+
+            var environmentInstance = await GetEnvironmentAsync(_projectName, environmentName);
+           
+            agentSettings.EnvironmentId = environmentInstance.Id;
+            agentSettings.ProjectName = environmentInstance.Project.Name;
+            agentSettings.ProjectId = environmentInstance.Project.Id.ToString();            
+            Trace.Info("vm resource configuration: environment id: '{0}', project name: '{1}', project id: '{2}'", agentSettings.EnvironmentId, agentSettings.ProjectName, agentSettings.ProjectId);
+        }
+
+        public override string GetFailedToFindPoolErrorString() => StringUtil.Loc("FailedToFindEnvironment");
```

Comment:
is it worth including the environment name in this error string?

Reply:
yes, this will hep us in future to differentiate the error message.